src/services/data_model_engine.js

```diff
@@ -220,6 +221,24 @@ export default class DataModelEngine {
     return summary;
   }
 
+  async uploadBundle(bundleId, storagePeriods) {
```

Comment:
The backoff mechanism should be extracted, generalised, and tested on its own (using sinons fake timers for example). This class should just use this mechanism. The code will be cleaner and easier to test this way.